File: Data_Structures/javascript/BinaryTree/preOrderTraversal.js
const PreOrderTraverse = (node) => {
   if (node !== null) {
     //Print the node
     console.log(node.key);

     //Traverse the left subtree
     PreOrderTraverse(node.left);

     //Traverse the right subtree
     PreOrderTraverse(node.right);
   }
};

const tree = new BinarySearchTree();
tree.insert(11);
tree.insert(9);
tree.insert(15);
tree.insert(8);
tree.insert(10);
tree.insert(12);
tree.insert(16);

const root = tree.getRoot();
PreOrderTraverse(root);

File: Data_Structures/C/BinarySearchTree/preOrderTraversal.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    char data;
    struct Node *left;
    struct Node *right;
};

void    preOrder(struct Node *root)
{
    if (!root) return;
    printf("%c ", root->data);
    preOrder(root->left);
    preOrder(root->right);
}

struct Node     *newNode(int data)
{
    struct Node *new_node = malloc(sizeof(struct Node));
    new_node->data = data;
    new_node->left = NULL;
    new_node->right = NULL;
    return new_node;
}

// Function to Insert Node in a Binary Search Tree
struct Node    *Insert(struct Node *root, char data) {
	if (!root)
        root = newNode(data);
	else if (data <= root->data)
		root->left = Insert(root->left,data);
	else
		root->right = Insert(root->right,data);
	return root;
}

int     main() {
/*Code To Test the logic

Creating an example tree
	    M
	   / \
	  B   Q
	 / \   \
	A   C   Z
*/
	struct Node *root = NULL;
	root = Insert(root,'M'); root = Insert(root,'B');
	root = Insert(root,'Q'); root = Insert(root,'Z');
	root = Insert(root,'A'); root = Insert(root,'C');
	//Print Nodes in Preorder.
	printf("Preorder: ");
	preOrder(root);
	printf("\n");
	return (0);
}
